### ISSUE TITLE
How to create exit signal by trade (use information bounded on trade) ?

### ISSUE THREAD

**Issue by eromoe** [freqtrade/freqtrade#8335]
## Describe your environment

  * Operating system:  win 10
  * Python Version: 3.8
  * CCXT version: ccxt==2.9.4
  * Freqtrade Version: freqtrade 2023.3.dev-5d9868b54
  * 
## Your question

How to create exit signal on trade ?
For example, I build a model which label by triple barrier method: http://www.sefidian.com/2021/06/26/labeling-financial-data-for-machine-learning/#:~:text=The%20triple%20barrier%20method%20involves,taking%20or%20stop%2Dloss%20limits.

Each trade with triple barrier I need bind data: 
- volatility ( base on trade open time,  look back 100 tick,  caclulate rolling std or ewmstd ) 
- vertical bar length ( 30 minute)

Exit Rules:
loop and check exitsting trades 
- tp =  current_open -  trade_open > volatility 
- sl =  current_open -  trade_open < -volatility 
- vbar = current_time - trade_open_time > 30

From docs https://www.freqtrade.io/en/stable/strategy-customization/#stoploss_from_open ,
- `custom_stoploss` only return the stoploss rate , can not fit my purpose . 
- `custom_exit` can access trades, but can not bind data upon.  would like have a `custom_entry` to bind data on trade object.

 Could you give me some advise ? I would happy to contribute a demo of triple barrier exit .

**Comment by eromoe** [comment]
Trying with `custom_stoploss` ,  is this a correct /good way ?
```python
    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:
        
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        if dataframe.empty:
            return 10
        
        trade_date = timeframe_to_prev_date(self.timeframe, trade.open_date_utc)
        trade_candle = dataframe.loc[dataframe['date'] == trade_date]
        
        # sell if hit vertical barrier
        if datetime.now() > (trade_date + pd.Timedelta(minutes=30)):
            return 0
        
        # sell if hit upper / lower barrier
        return stoploss_from_absolute(trade_candle['vol'] / trade_candle['open'] , current_rate, is_short=trade.is_short)
```

**Comment by froggleston** [comment]
I'm not sure what "bind data to trade object" means, but the Trade object is immutable - you cannot add or change data associated with the trade itself. You have access to any dataframe imformation, or custom information held in any class member variables you might have created, e.g. self.custom_info

Yes, using trade_candle in this way is correct, but this won't work if the trade is a long way back in time (exchange dependent, but anywhere from 300 candles back). If your trade durations are relatively short, then this will work.

However datetime.now() is not backtestable - use the current_time argument. Also, returning 0 from this function will result in very tight stoploss calculations and will likely result in problematic backtesting. If you want to exit immediately, use custom_exit.

There is a fixed TP/SL strategy example in the freqtrade strategies repository that might help: https://github.com/freqtrade/freqtrade-strategies/blob/main/user_data/strategies/FixedRiskRewardLoss.py

**Comment by xmatthias** [comment]
`return 0` from custom_stoploss is considered an invalid stoploss - which will be ignored (including a warning message in the logs).

**Comment by eromoe** [comment]
@froggleston I am confused on https://github.com/freqtrade/freqtrade-strategies/blob/main/user_data/strategies/FixedRiskRewardLoss.py#L80 , in most case `break_even_profit_pct < take_profit_pct`  , so only `if (current_profit >= break_even_profit_pct)` would trigger, never hit take_profit_pct.  Do I misunderstand ?

**Comment by eromoe** [comment]
Rewrite by `custom_exit`  , 

```python
    def custom_exit(self, pair: str, trade: 'Trade', current_time: 'datetime', current_rate: float,
                    current_profit: float, **kwargs):
        
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        trade_date = timeframe_to_prev_date(self.timeframe, trade.open_date_utc)
        trade_candle = dataframe.loc[lambda df: df['date'] == trade_date]
        
        # trade.open_rate is open_price

        if trade_candle.empty:
            return None
        
        if current_time > (trade_date + pd.Timedelta(minutes=30)):
            return 'hit_vertical'
        
        last_candle = dataframe.iloc[-1].squeeze()

        if last_candle['close'] > (trade_candle['open'] + trade_candle['vol']):
            return 'hit_upper'

        if last_candle['close'] < (trade_candle['open'] - trade_candle['vol']):
            return 'hit_lower'
```

I also wonder should I divide `hit_upper` and `hit_lower` mechanism  in `custom_stoploss` ?

**Comment by xmatthias** [comment]
> @froggleston I am confused on https://github.com/freqtrade/freqtrade-strategies/blob/main/user_data/strategies/FixedRiskRewardLoss.py#L80 , in most case `break_even_profit_pct < take_profit_pct` , so only `if (current_profit >= break_even_profit_pct)` would trigger, never hit take_profit_pct. Do I misunderstand ?

what "triggers" isn't relevant - that sample is in `custom_stoploss` - not `custom_exit`.
the calculation of what's returned is quite different - the idea being:
* once reaching break-even offset -> move stop to break even
* once reaching take-profit -> move stop closer to current price.

**Comment by eromoe** [comment]
@xmatthias Thank you ! Understand now. 
I found I misunderstood `xxx_rate` as `xxx_percentage` which actually stand for `xxx_price` , no wonder this code feeled weird to me .

**Comment by eromoe** [comment]
Thanks, but I am still a bit confused , 

For example,  

# Case 1,  current_profit is between  take_profit_pct and break_even_profit_pct

```python
import pandas as pd
import numpy as np
import plotly.graph_objs as go
import plotly.express as px

risk_distance = 0.05
open_rate = np.arange(10, 15, 0.5)
current_rate = np.arange(10, 15, 0.5) +0.1

current_profit = current_rate / open_rate -1
take_profit_pct = (open_rate + 3.5*risk_distance)/open_rate -1
break_even_profit_pct = (current_rate + 1.0*risk_distance)/current_rate -1

x = np.arange(len(open_rate))
fig = go.Figure()
fig.add_trace(
    go.Line(x=x, y=current_profit, name='current_profit')
)
fig.add_trace(
    go.Line(x=x, y=take_profit_pct, name='take_profit_pct')
)
fig.add_trace(
    go.Line(x=x, y=break_even_profit_pct, name='break_even_profit_pct')
)

fig.layout.update(dict(width=800))
fig.show()

```

![image](https://user-images.githubusercontent.com/3938751/225594308-9a8cbdde-de1d-4b58-b057-40d5fa06a48e.png)

`break_even_profit_pct`  is always below `take_profit_pct` ,  final stoploss always take `break_even_profit_pct`

# Case 2 ,  current_profit is above both take_profit_pct and break_even_profit_pct

```
risk_distance = 0.05
open_rate = np.arange(10, 15, 0.5)
current_rate = np.arange(10, 15, 0.5) +1

current_profit = current_rate / open_rate -1
take_profit_pct = (open_rate + 3.5*risk_distance)/open_rate -1
break_even_profit_pct = (current_rate + 1.0*risk_distance)/current_rate -1

x = np.arange(len(open_rate))
fig = go.Figure()
fig.add_trace(
    go.Line(x=x, y=current_profit, name='current_profit')
)
fig.add_trace(
    go.Line(x=x, y=take_profit_pct, name='take_profit_pct')
)
fig.add_trace(
    go.Line(x=x, y=break_even_profit_pct, name='break_even_profit_pct')
)

fig.layout.update(dict(width=800))
fig.show()
```

![image](https://user-images.githubusercontent.com/3938751/225594937-1501b409-b54f-4648-9b22-fd89e1d9028b.png)

---

As above , only case 2 `current_profit` is over `take_profile_pct`, the final stoploss could become `take_profile_pct` . 
But `current_profit`  usually need go through case 1 , then reach to case 2 , unless there is a quick price upward. 
That's what I mean it may never hit `take_profit_pct` , always use `break_even_profit_pct ` , which looks like a useless strategy (forgive my ignorance ) .

**Comment by froggleston** [comment]
open_rate needs to be the fixed trade open rate, not each candle open rate, if I'm reading your code correctly.

**Comment by eromoe** [comment]
@froggleston You can consider the open_rate beyond to 10 trades. 
My concern is the strategy would always hit `break_even_profit_pct` at first  while current_rate is greater than open_rate .

----

I can simulate one trade with current_rate growing up: 

```python
risk_distance = 0.05
open_rate = np.ones(10)*10
current_rate = np.arange(10, 11, 0.1) +0.1

current_profit = current_rate / open_rate -1
take_profit_pct = (open_rate + 3.5*risk_distance)/open_rate -1
break_even_profit_pct = (current_rate + 1.0*risk_distance)/current_rate -1
```

![image](https://user-images.githubusercontent.com/3938751/225791018-73a4f6be-6456-48e8-8a13-2669cb0f7868.png)

That's why I said it go through case 1 (place break_even)  , then reach to case 2 ( plase trake_profit)

**Comment by xmatthias** [comment]
i don't really see your problem/point, really.
we're not talking about a `custom_exit()` - but about `custom_stoploss()` - at least from the example you linked to. You can modify stoploss 20 times in different ways (as long as you pull it closer over time). 

so yes, break_even_offset will definitely be hit first - before then hitting the stoploss.

for EVERY one of your samples above - think the following (we entered the long trade at 100$):
* initial stop is -20% (80$)
* once you reach 5% profit (price is ~105$), move stoploss to break even (say - the open rate - ~-5% - or 100.1$)
* once you reach 10% (price reaches 110$ - your take-profit target) - you move the stoploss to 2% below current price - in this case ~108$.

Your're still in the trade (also beyond 110$) - so your trade could really go to 120$ - or beyond that - with the stoploss trailing price with a 2% offset. The stoploss will only go up, never down, so once the trade reaches a pullback point (of more than 2%) - your stop will trigger, and you'll exit the trade.


You can play this with percentages, or absolute numbers - but i don't really see why hitting "break even" first would be a problem ... that's a mathematical fact, which will hold true for every trade - unless your goal is to make losses, and your "take profit" (now in quotes!) is negative.

**Comment by eromoe** [comment]
Oh, I got it! My bad, I did really mix up the custom_exit and custom_stoploss , also mix up stoploss and takeprofit  (in takeprofit view,  break_even_offset would hit and sell, but actually may not, can keep go up ). Before, I thought the `custom_stoploss ` is used for controling the upper_pct(takeprofit)  and lower_pct(stoploss) . But actually there is always and only need set stoploss ,  as your said `10% (price reaches 110$ - your take-profit target) - you move the stoploss to 2%` ,  takeprofit is achieved by placing a stoploss. 

Thank you for kindness explanation , I would try to make some workable strategies to contribute freqtrade community :)
